### ISSUE TITLE
regex hangs on match start

### ISSUE THREAD

**Issue by Inc0n** [asciian/regex#1]
trying regex string "^+" or "^\+" to match  '+' at start would hang
```
CL-USER> (re:scan "^+" "abcd")
(#<CLOSURE (LAMBDA (REGEX.CORE::DATA REGEX.CORE::REG REGEX.CORE::CONT)
             :IN
             REGEX.CORE:RE/REPEAT) {100406F20B}>
 #<FUNCTION (LAMBDA (REGEX.CORE::DATA REGEX.CORE::REG REGEX.CORE::CONT)
              :IN
              REGEX.CORE:RE/OK) {52DE956B}>)
```
but this would work
```
CL-USER> (re:scan "^[+]" "abcd")
             ....
             :IN
             REGEX.CORE:RE/BRACKETS) {100250091B}>
 #<FUNCTION (LAMBDA (REGEX.CORE::DATA REGEX.CORE::REG REGEX.CORE::CONT)
              :IN
              REGEX.CORE:RE/OK) {52DE814B}>)
NIL
```

**Comment by tani** [comment]
Thank you for the issue. However, `^+` is an invalid expression of POSIX Extended Reuglar Exression. You could check it with grep like this.

```sh
$ grep -E '^+' <(echo '+')
grep: repetition-operator operand invalid
```

We follow the syntax of grep with `-E` option that is the Extended Reuglar Exression.
Cheers.

**Comment by Inc0n** [comment]
I think the mistake is that the regex should be “^\+”, should some type of check be implemented into the syntax check? For this type of issue.

**Comment by tani** [comment]
Yes, I agree that we should implemented into the syntax check, then the better way is reimplementing the parser with full-futured parser generators such as `esrap` because we could expect that these parser generators can inform errors and suggestions to correct the syntax.

Of couse we could add  the validator for these errors with this engine but the syntax is not in the regular grammar, thus the validator is incomplete to check all errors.
